File: src/lib/wow/db/caches/db-creatures.ts
/**
 * Creature display cache, ported from wow.export (src/js/db/caches/DBCreatures.js).
 */
import { write } from '@/lib/wow/log';

import { doOnce } from '../../formats/generics';
import { WDCReader } from '../wdc-reader';

/** A creature/item display entry mapped to a model fileDataID. */
export interface ModelDisplay {
  ID: number;
  modelID?: number;
  textures: number[];
  extraGeosets?: number[];
}

const creatureDisplays = new Map<number, ModelDisplay[]>();
const displayIDToFileDataID = new Map<number, number>();
let isInitialized = false;

/** Initialize creature data. */
export const initializeCreatureData = doOnce('initializeCreatureData', async () => {
  if (isInitialized) return;

  write('Loading creature textures...');

  const creatureDisplayInfo = new WDCReader('DBFilesClient/CreatureDisplayInfo.db2');
  await creatureDisplayInfo.parse();

  const creatureModelData = new WDCReader('DBFilesClient/CreatureModelData.db2');
  await creatureModelData.parse();

  const creatureGeosetMap = new Map<number, number[]>();

  const creatureDisplayInfoGeosetData = new WDCReader('DBFilesClient/CreatureDisplayInfoGeosetData.db2');
  await creatureDisplayInfoGeosetData.parse();
  // CreatureDisplayInfoID => Array of geosets to enable which should only be used if CreatureModelData.CreatureDisplayInfoGeosetData != 0
  for (const geosetRow of creatureDisplayInfoGeosetData.getAllRows().values()) {
    const displayInfoID = geosetRow.CreatureDisplayInfoID as number;
    if (!creatureGeosetMap.has(displayInfoID)) creatureGeosetMap.set(displayInfoID, []);

    creatureGeosetMap.get(displayInfoID)!.push(((geosetRow.GeosetIndex as number) + 1) * 100 + (geosetRow.GeosetValue as number));
  }

  const creatureDisplayInfoMap = new Map<number, ModelDisplay>();
  const modelIDToDisplayInfoMap = new Map<number, number[]>();

  // Map all available texture fileDataIDs to model IDs.
  for (const [displayID, displayRow] of creatureDisplayInfo.getAllRows()) {
    const modelID = displayRow.ModelID as number;
    creatureDisplayInfoMap.set(displayID, {
      ID: displayID,
      modelID,
      textures: (displayRow.TextureVariationFileDataID as number[]).filter((e) => e > 0),
    });

    if (modelIDToDisplayInfoMap.has(modelID)) modelIDToDisplayInfoMap.get(modelID)!.push(displayID);
    else modelIDToDisplayInfoMap.set(modelID, [displayID]);
  }

  // Using the texture mapping, map all model fileDataIDs to used textures.
  for (const [modelID, modelRow] of creatureModelData.getAllRows()) {
    if (modelIDToDisplayInfoMap.has(modelID)) {
      const fileDataID = modelRow.FileDataID as number;
      const displayIDs = modelIDToDisplayInfoMap.get(modelID)!;
      const modelIDHasExtraGeosets = (modelRow.CreatureGeosetDataID as number) > 0;

      for (const displayID of displayIDs) {
        displayIDToFileDataID.set(displayID, fileDataID);

        const display = creatureDisplayInfoMap.get(displayID)!;

        if (modelIDHasExtraGeosets) {
          display.extraGeosets = [];
          if (creatureGeosetMap.has(displayID)) display.extraGeosets = creatureGeosetMap.get(displayID)!;
        }

        if (creatureDisplays.has(fileDataID)) creatureDisplays.get(fileDataID)!.push(display);
        else creatureDisplays.set(fileDataID, [display]);
      }
    }
  }

  write('Loaded textures for %d creatures', creatureDisplays.size);
  isInitialized = true;
});

/** Gets creature skins from a given file data ID. */
export function getCreatureDisplaysByFileDataID(fileDataID: number): ModelDisplay[] | undefined {
  return creatureDisplays.get(fileDataID);
}

/** Gets the file data ID for a given display ID. */
export function getFileDataIDByDisplayID(displayID: number): number | undefined {
  return displayIDToFileDataID.get(displayID);
}

export function getCreatureCacheStats(): { initialized: boolean; creatureDisplays: number; displayIDToFileDataID: number } {
  return {
    initialized: isInitialized,
    creatureDisplays: creatureDisplays.size,
    displayIDToFileDataID: displayIDToFileDataID.size,
  };
}

export default { initializeCreatureData, getCreatureDisplaysByFileDataID, getFileDataIDByDisplayID };

File: golang/internal/wow/db/caches/db_creatures.go
package caches

import (
	"context"
	"sync"

	"github.com/pqhuy98/wow-converter/internal/wow/db"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

// ModelDisplay is a creature/item display entry mapped to a model fileDataID.
type ModelDisplay struct {
	ID           uint32
	ModelID      uint32
	Textures     []uint32
	ExtraGeosets []uint32
}

var (
	creatureDisplays       = make(map[uint32][]ModelDisplay)
	displayIDToFileDataID  = make(map[uint32]uint32)
	creatureInitialized    bool
	creatureOnce           sync.Once
	creatureErr            error
)

// InitializeCreatureData loads creature display tables.
func InitializeCreatureData(ctx context.Context) error {
	creatureOnce.Do(func() {
		if creatureInitialized {
			return
		}
		log.Write("Loading creature textures...")

		displayInfo := db.NewWDCReader("DBFilesClient/CreatureDisplayInfo.db2", nil)
		if err := displayInfo.Parse(ctx, nil); err != nil {
			creatureErr = err
			return
		}
		modelData := db.NewWDCReader("DBFilesClient/CreatureModelData.db2", nil)
		if err := modelData.Parse(ctx, nil); err != nil {
			creatureErr = err
			return
		}
		geosetData := db.NewWDCReader("DBFilesClient/CreatureDisplayInfoGeosetData.db2", nil)
		if err := geosetData.Parse(ctx, nil); err != nil {
			creatureErr = err
			return
		}

		creatureGeosetMap := make(map[uint32][]uint32)
		for _, geosetRow := range geosetData.GetAllRows() {
			displayInfoID := toUint32(geosetRow["CreatureDisplayInfoID"])
			geoset := (toUint32(geosetRow["GeosetIndex"])+1)*100 + toUint32(geosetRow["GeosetValue"])
			creatureGeosetMap[displayInfoID] = append(creatureGeosetMap[displayInfoID], geoset)
		}

		creatureDisplayInfoMap := make(map[uint32]ModelDisplay)
		modelIDToDisplayInfoMap := make(map[uint32][]uint32)

		for displayID, displayRow := range displayInfo.GetAllRows() {
			modelID := toUint32(displayRow["ModelID"])
			textures := filterPositive(toUint32Slice(displayRow["TextureVariationFileDataID"]))
			creatureDisplayInfoMap[displayID] = ModelDisplay{
				ID: displayID, ModelID: modelID, Textures: textures,
			}
			modelIDToDisplayInfoMap[modelID] = append(modelIDToDisplayInfoMap[modelID], displayID)
		}

		for modelID, modelRow := range modelData.GetAllRows() {
			displayIDs, ok := modelIDToDisplayInfoMap[modelID]
			if !ok {
				continue
			}
			fileDataID := toUint32(modelRow["FileDataID"])
			modelIDHasExtraGeosets := toUint32(modelRow["CreatureGeosetDataID"]) > 0
			for _, displayID := range displayIDs {
				displayIDToFileDataID[displayID] = fileDataID
				display := creatureDisplayInfoMap[displayID]
				if modelIDHasExtraGeosets {
					display.ExtraGeosets = creatureGeosetMap[displayID]
				}
				creatureDisplays[fileDataID] = append(creatureDisplays[fileDataID], display)
			}
		}

		log.Write("Loaded textures for %d creatures", len(creatureDisplays))
		creatureInitialized = true
	})
	return creatureErr
}

// GetCreatureDisplaysByFileDataID returns creature skins for a file data ID.
func GetCreatureDisplaysByFileDataID(fileDataID uint32) []ModelDisplay {
	return creatureDisplays[fileDataID]
}

// GetFileDataIDByDisplayID returns the file data ID for a display ID.
func GetFileDataIDByDisplayID(displayID uint32) (uint32, bool) {
	id, ok := displayIDToFileDataID[displayID]
	return id, ok
}

// GetCreatureCacheStats returns cache stats.
func GetCreatureCacheStats() (initialized bool, creatureDisplaysCount, displayIDMap int) {
	return creatureInitialized, len(creatureDisplays), len(displayIDToFileDataID)
}

// ResetCreatureCache clears creature cache.
func ResetCreatureCache() {
	creatureDisplays = make(map[uint32][]ModelDisplay)
	displayIDToFileDataID = make(map[uint32]uint32)
	creatureInitialized = false
	creatureOnce = sync.Once{}
	creatureErr = nil
}

func toUint32Slice(v any) []uint32 {
	switch x := v.(type) {
	case []uint32:
		return x
	case []int64:
		out := make([]uint32, len(x))
		for i, n := range x {
			out[i] = uint32(n)
		}
		return out
	case []int32:
		out := make([]uint32, len(x))
		for i, n := range x {
			out[i] = uint32(n)
		}
		return out
	default:
		return nil
	}
}

func filterPositive(ids []uint32) []uint32 {
	out := make([]uint32, 0, len(ids))
	for _, id := range ids {
		if id > 0 {
			out = append(out, id)
		}
	}
	return out
}
